### PR TITLE
adding prefix to all resources as capx- and enable tilt

### PR DESCRIPTION
**Pull request by deepakm-ntnx**
This will make prefix of all resource names to capx-

Testing steps followed:
<pre>
make docker-build
</pre>

<pre>
#~ make deploy
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/controller-gen "crd:crdVersions=v1" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
mkdir -p ~/.cluster-api/overrides/infrastructure-nutanix/v0.0.0
cd config/manager && /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/kustomize edit set image controller=ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller:latest
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/kustomize build config/default > ~/.cluster-api/overrides/infrastructure-nutanix/v0.0.0/infrastructure-components.yaml
cp ./metadata.yaml ~/.cluster-api/overrides/infrastructure-nutanix/v0.0.0
cp ./cluster-template.yaml ~/.cluster-api/overrides/infrastructure-nutanix/v0.0.0
cp ./clusterctl.yaml ~/.cluster-api/clusterctl.yaml
rm -f /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/conversion-gen
Downloading k8s.io/code-generator/cmd/conversion-gen@v0.23.6
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/conversion-gen \
	--input-dirs=./api/v1alpha4 \
	--input-dirs=./api/v1beta1 \
	--build-tag=ignore_autogenerated_core \
	--output-file-base=zz_generated.conversion \
	--go-header-file=./hack/boilerplate.go.txt
go fmt ./...
go vet ./...
KUBEBUILDER_ASSETS="/Users/deepak.muley/Library/Application Support/io.kubebuilder.envtest/k8s/1.23.5-darwin-amd64" go test ./... -coverprofile cover.out
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix	[no test files]
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/api/v1alpha4	[no test files]
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/api/v1beta1	[no test files]
ok  	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/controllers	11.158s	coverage: 2.3% of statements
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pkg/client	[no test files]
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pkg/context	[no test files]
ok  	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pkg/nutanix	1.349s	coverage: 62.0% of statements
ok  	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pkg/nutanix/v3	1.216s	coverage: 58.1% of statements
?   	github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pkg/utils	[no test files]
GOOS=linux GOARCH=amd64 KO_DOCKER_REPO=ko.local /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/ko build -B -t latest -L .
2022/06/04 11:29:18 Using base gcr.io/distroless/static:nonroot@sha256:2556293984c5738fc75208cce52cf0a4762c709cf38e4bf8def65a61992da0ad for github.com/nutanix-cloud-native/cluster-api-provider-nutanix
2022/06/04 11:29:18 Building github.com/nutanix-cloud-native/cluster-api-provider-nutanix for linux/amd64
2022/06/04 11:29:26 Loading ko.local/cluster-api-provider-nutanix:2679a89ad3c740ffb9c040115f4896c8423872012493c96fb2107f3b8acdfeb2
2022/06/04 11:29:27 Loaded ko.local/cluster-api-provider-nutanix:2679a89ad3c740ffb9c040115f4896c8423872012493c96fb2107f3b8acdfeb2
2022/06/04 11:29:27 Adding tag latest
2022/06/04 11:29:27 Added tag latest
ko.local/cluster-api-provider-nutanix:2679a89ad3c740ffb9c040115f4896c8423872012493c96fb2107f3b8acdfeb2
docker tag ko.local/cluster-api-provider-nutanix:latest ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller:latest
kind load docker-image --name capi-test ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller:latest
Image: "ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller:latest" with ID "sha256:8f5eb48f8137f269a988f5e3e6fc9b33081dd38eb33ccc40dac98bcb0e0b53af" found to be already present on all nodes.
clusterctl init --infrastructure nutanix:v0.0.0 -v 9
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
Fetching providers
Using Override="infrastructure-components.yaml" Provider="infrastructure-nutanix" Version="v0.0.0"
Fetching File="metadata.yaml" Provider="cluster-api" Type="CoreProvider" Version="v1.1.4"
Using Override="metadata.yaml" Provider="infrastructure-nutanix" Version="v0.0.0"
Creating Namespace="cert-manager-test"
Creating Issuer="test-selfsigned" Namespace="cert-manager-test"
Creating Certificate="selfsigned-cert" Namespace="cert-manager-test"
Deleting Namespace="cert-manager-test"
Deleting Issuer="test-selfsigned" Namespace="cert-manager-test"
Deleting Certificate="selfsigned-cert" Namespace="cert-manager-test"
Skipping installing cert-manager as it is already installed
Installing Provider="infrastructure-nutanix" Version="v0.0.0" TargetNamespace="capx-system"
Creating objects Provider="infrastructure-nutanix" Version="v0.0.0" TargetNamespace="capx-system"
Creating Namespace="capx-system"
Creating CustomResourceDefinition="nutanixclusters.infrastructure.cluster.x-k8s.io"
I0604 11:29:39.091718   77127 request.go:665] Waited for 1.019566239s due to client-side throttling, not priority and fairness, request: GET:https://127.0.0.1:49830/apis/autoscaling/v2beta2?timeout=30s
Creating CustomResourceDefinition="nutanixmachines.infrastructure.cluster.x-k8s.io"
Creating CustomResourceDefinition="nutanixmachinetemplates.infrastructure.cluster.x-k8s.io"
Creating ServiceAccount="capx-controller-manager" Namespace="capx-system"
Creating Role="capx-leader-election-role" Namespace="capx-system"
Creating ClusterRole="capx-system-capx-manager-role"
Creating ClusterRole="capx-system-capx-metrics-reader"
Creating ClusterRole="capx-system-capx-proxy-role"
Creating RoleBinding="capx-leader-election-rolebinding" Namespace="capx-system"
Creating ClusterRoleBinding="capx-system-capx-manager-rolebinding"
Creating ClusterRoleBinding="capx-system-capx-proxy-rolebinding"
Creating ConfigMap="capx-manager-config" Namespace="capx-system"
Creating Secret="capx-nutanix-creds" Namespace="capx-system"
Creating Service="capx-controller-manager-metrics-service" Namespace="capx-system"
Creating Deployment="capx-controller-manager" Namespace="capx-system"
Creating inventory entry Provider="infrastructure-nutanix" Version="v0.0.0" TargetNamespace="capx-system"
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
# cd config/manager && /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/kustomize edit set image controller=ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller:latest
# /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/bin/kustomize build config/default | kubectl apply -f -
</pre>

<pre>
kubectl -n capx-system get pods
NAME                                      READY   STATUS    RESTARTS   AGE
capx-controller-manager-5854bf644-lgz75   2/2     Running   1          11m
</pre>

Added tilt support
To test following steps below,
- create a kind cluster if does not exist already
- install tilt
- download cluster-api repo
- create following tilt-settings.json
<pre>
{
  "default_registry": "ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller",
  "provider_repos": [
    "../cluster-api-provider-nutanix"
  ],
  "enable_providers": [
    "nutanix",
    "docker",
    "kubeadm-bootstrap",
    "kubeadm-control-plane"
  ],
  "kind_cluster_name": "tilt-test-cluster" <===provider your cluster
name if not capi-test which is default name
}
</pre>
- run "tilt up"

### CODE REVIEW

**Review by tuxtof — CHANGES_REQUESTED** [review]
ok on the idea but we need to clean some name to avoid redundant name like

capx-nutanix-creds

**Comment by tuxtof** [comment]
@deepakm-ntnx can you clean the creds name please

**Comment by tuxtof** [comment]
@deepakm-ntnx any update on this one ?

**Review by tuxtof — CHANGES_REQUESTED** [review]

File: tilt-provider.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply tilt-provider.json:1]: its suggested by cluster-api itself https://cluster-api.sigs.k8s.io/developer/tilt.html#developing-cluster-api-with-tilt

for a simplified workflow that offers easy deployments and rapid iterative builds.

[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply tilt-provider.json:1]: also, many other providers also ahve this file including cluster-api repo itself and vsphere etc

File: docs/getting_started.md [hunk 2]
```diff
@@ -28,22 +28,27 @@ clusterctl generate cluster mycluster -i nutanix --list-variables
...
   - NUTANIX_MACHINE_TEMPLATE_IMAGE_NAME
+  - NUTANIX_PASSWORD
   - NUTANIX_PRISM_ELEMENT_CLUSTER_NAME
   - NUTANIX_SSH_AUTHORIZED_KEY
   - NUTANIX_SUBNET_NAME
+  - NUTANIX_USER
 
 Optional Variables:
-  - CLUSTER_NAME
+  - CLUSTER_NAME                     (defaults to mycluster)
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply docs/getting_started.md:40]: this is the exact output from list-variables. why do you think its not supported? are you saying clusterctl mandates cluster name on cli even if its set?

File: config/manager/manager.yaml [hunk 3]
```diff
@@ -32,8 +32,8 @@ spec:
       labels:
         control-plane: controller-manager
     spec:
-      securityContext:
-        runAsNonRoot: true
+      # securityContext:
```
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply config/manager/manager.yaml:35]: Agree with you but for following reason, have temporarily commented it out
1. we already use nonroot image as follows
<pre>
2022/05/27 15:02:00 Using base gcr.io/distroless/static:nonroot@sha256:2556293984c5738fc75208cce52cf0a4762c709cf38e4bf8def65a61992da0ad for github.com/nutanix-cloud-native/cluster-api-provider-nutanix
</pre>
2. when i was running tilt, i was getting following error 
<pre>
  Normal   Scheduled  35s               default-scheduler  Successfully assigned capx-system/capx-controller-manager-6586d6f577-tqbzf to tilt-test-cluster-control-plane
  Normal   Pulling    33s               kubelet            Pulling image "gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0"
  Normal   Pulled     21s               kubelet            Successfully pulled image "gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0" in 11.737437542s
  Normal   Created    21s               kubelet            Created container kube-rbac-proxy
  Normal   Started    21s               kubelet            Started container kube-rbac-proxy
  Normal   Pulled     6s (x4 over 21s)  kubelet            Container image "ghcr.io/nutanix-cloud-native/cluster-api-provider-nutanix/controller/ghcr.io_nutanix-cloud-native_cluster-api-provider-nutanix_controller:tilt-5cf0bb2c082e5dcd" already present on machine
  Warning  Failed     6s (x4 over 21s)  kubelet            Error: container has runAsNonRoot and image will run as root (pod: "capx-controller-manager-6586d6f577-tqbzf_capx-system(913940f1-720d-49b7-ad82-c3e6959be4f4)", container: manager) <====
</pre>

will explore more if we can set some setting in tilt. i didnt find this set in other providers as well as a data point.

File: docs/getting_started.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tuxtof) [reply docs/getting_started.md:40]: yes never use CLUSTER_NAME ENV variable , you always need to specify it on command line

**Review by tuxtof — APPROVED** [review]

**Review by yannickstruyf3 — APPROVED** [review]

File: config/manager/manager.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (deepakm-ntnx) [reply config/manager/manager.yaml:35]: we are using nonRoot user inside container so technically we are. will file a tracker to make tilt work with this setting as well

**Pull request merged**